File: C++/13. Roman to Integer.cpp
#include <unordered_map>
#include <string>

using namespace std;

int get_val(char c)
{
    switch (c)
    {
    case 'I':
        return 1;
    case 'V':
        return 5;
    case 'X':
        return 10;
    case 'L':
        return 50;
    case 'C':
        return 100;
    case 'D':
        return 500;
    case 'M':
        return 1000;
    }
    return 0;
}

int romanToInt(string s)
{
    int result = 0;

    for (int i = 0; i < s.size(); i++)
    {
        if (s[i] == 'I' and (i + 1 < s.size()) and (s[i + 1] == 'V' or s[i + 1] == 'X'))
            result -= 1;
        else if (s[i] == 'X' and (i + 1 < s.size()) and (s[i + 1] == 'L' or s[i + 1] == 'C'))
            result -= 10;
        else if (s[i] == 'C' and (i + 1 < s.size()) and (s[i + 1] == 'D' or s[i + 1] == 'M'))
            result -= 100;
        else
            result += get_val(s[i]);
    }

    return result;
}

File: Java/13. Roman to Integer.java
class Solution {
    private int getRomanValue(char c) {
        return switch (c) {
            case 'I' -> 1;
            case 'V' -> 5;
            case 'X' -> 10;
            case 'L' -> 50;
            case 'C' -> 100;
            case 'D' -> 500;
            case 'M' -> 1000;
            default -> 0;
        };
    }

    public int romanToInt(String s) {
        int result = 0;
        int n = s.length();
        char c;
        for (int i = 0; i < n; i++) {
            c = s.charAt(i);
            if (c == 'I' && i < n - 1 && (s.charAt(i + 1) == 'V' || s.charAt(i + 1) == 'X'))
                result -= 1;
            else if (c == 'X' && i < n - 1 && (s.charAt(i + 1) == 'L' || s.charAt(i + 1) == 'C'))
                result -= 10;
            else if (c == 'C' && i < n - 1 && (s.charAt(i + 1) == 'D' || s.charAt(i + 1) == 'M'))
                result -= 1;
            else
                result += getRomanValue(c);
        }

        return result;
    }
}

/*
 * 
 * I 1
 * IV 4
 * IX 9
 * V 5
 * X 10
 * XL 40
 * XC 90
 * L 50
 * C 100
 * CD 400
 * CM 900
 * D 500
 * M 1000
 * 
 */